bfieldtools/sensors.py

```diff
@@ -0,0 +1,310 @@
+        """
+        Finalize the geometric information of the sensor
+        and transform it using 'transform'
+        
+        After finalize the object is ready for field measurements
+        and reciprocal field calculations
+
+        Parameters
+        ----------
+        transform : ndarray (4,4) 
+           transformation matrix
+
+
+        Returns
+        -------
+        None
+
+        """
+        self.base_sensor = False
+        t = transform
+        if not isinstance(t, np.ndarray):
+            raise TypeError("transform must be ndarray")
+        if not t.shape == (4, 4):
+            raise ValueError("transform shape must be (4,4)")
+
+    @abstractmethod
+    def update_position(self, transform):
+        """
+        Update the geometric information of the sensor
+        by transform
+        
+        transform acts on the initial coordinates of the sensor
+        (not on the transformed coordinates)
+
+        Parameters
+        ----------
+        transform : ndarray (4,4) 
+           transformation matrix
+
+        Returns
+        -------
+        None
+
+        """
+        t = transform
+        if not isinstance(t, np.ndarray):
+            raise TypeError("transform must be ndarray")
+        if not t.shape == (4, 4):
+            raise ValueError("transform shape must be (4,4)")
+
+    @abstractmethod
+    def measure_bfield(self, bfield_func):
+        """
+        Magnetic field measurement
+
+        Parameters
+        ----------
+        bfield_func : function
+            function taking (N, 3) array of points 
+            as parameter and outputting (N, 3) array of magnetic field
+
+        Returns
+        -------
+        meas : float
+            Magnetic field measurement
+
+        """
+        pass
+
+    @abstractmethod
+    def measure_afield(self, afield_func):
+        """
+        Magnetic field measurement using vector potential
+
+        Parameters
+        ----------
+        afield_func : function
+            function that takes (N, 3) array of points 
+            as a parameter and outputs (N, 3) array of vector potential
+
+        Returns
+        -------
+        meas : float
+            Magnetic field measurement using vector potential
+
+        """
+        pass
+
+    @abstractmethod
+    def bfield_self(self, points):
+        """
+        Reciprocal B-field for the sensor
+
+        Parameters
+        ----------
+        points : ndarray (N, 3)
+            evaluation points
+
+        Returns
+        -------
+        B-field at the evaluation points
+
+        """
+        pass
+
+    @abstractmethod
+    def afield_self(self, points):
+        """
+        Reciprocal vector potential for the sensor
+
+        Parameters
+        ----------
+        points : ndarray (N, 3)
+            evaluation points
+
+        Returns
+        -------
+        B-field at the evaluation points
+
+        """
+        pass
+
+
+class MagnetometerLoop(BaseSensor):

```

Comment:
Should we perhaps implement loop-based computations as well as quadrature-point-based ones?

Reply:
Sure! For point-based, we should maybe write simple functions for the magnetic field and vector potential of magnetic dipoles in a separate file.

Reply:
I implemented functions for magnetic dipoles but did not include the quadrature-point based calculations yet. Maybe use those first for the opm implementation.